Refactor table operations structure in asset store

GiaJordan opened the pull request:
`make_synapse_table` a method that handled multiple table operations (creation, update, and replace) has been refactored into a `TableOperations` object with functionality separated into distinct methods. This will allow for easier implementation of new table operations in the future and makes the existing code easier to understand.

GiaJordan commented:
@GiaJordan update api endpoint

GiaJordan commented:
The api endpoint has been updated to accept the new parameter
@milen-sage @linglp

mialy-defelice commented:
@GiaJordan did you experience any issues calling to `validate` after you added this line:
```
@click.option(
    "-dt",
    "--data_type",
    callback=parse_comma_str_to_list,
    help=query_dict(model_commands, ("model", "validate", "data_type")),
)
```
When trying to run the command `schematic model --config config.yml validate -mp tests/data/1069/Bulk_rnaseq_filebased_example_error.model.csv` with `data_type` in the `config.yml` file, I get the following error stack: 
```
(schematicpy-py3.10) sh-3.2# schematic model --config config.yml validate -mp tests/data/1069/Bulk_rnaseq_filebased_example_error.model.csv
Starting schematic...
Traceback (most recent call last):
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/bin/schematic", line 6, in <module>
    sys.exit(main())
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1257, in invoke
    sub_ctx = cmd.make_context(cmd_name, args, parent=ctx)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 700, in make_context
    self.parse_args(ctx, args)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1048, in parse_args
    value, args = param.handle_parse_result(ctx, opts, args)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1630, in handle_parse_result
    value = invoke_param_callback(self.callback, ctx, self, value)
  File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 123, in invoke_param_callback
    return callback(ctx, param, value)
  File "/Users/mialydefelice/Documents/schematic_d/schematic/schematic/utils/cli_utils.py", line 165, in parse_comma_str_to_list
    return comma_string.split(",")
```
When I look what's happening, it looks like the `data_type` isn't being read in at all and `comma_string` here is empty, which is why the error is generated:
```
/Bulk_rnaseq_filebased_example_error.model.csv
Starting schematic...
> /Users/mialydefelice/Documents/schematic_d/schematic/schematic/utils/cli_utils.py(165)parse_comma_str_to_list()
-> return comma_string.split(",")
(Pdb) comma_string
(Pdb) param
<Option data_type>
```

If I add the `-dt` option from the command line it works:
```(schematicpy-py3.10) sh-3.2# schematic model --config config.yml validate -mp tests/data/1069/Bulk_rnaseq_filebased_example_error.model.csv -dt BulkRNA-seqAssay
Starting schematic...

error: For the attribute 'Sample ID', on row 3, '' is too short.
error: For the attribute 'Filename', on row 3, '' is too short.
[[3, 'Sample ID', "'' is too short", ''], [3, 'Filename', "'' is too short", '']]
```
Returning the expected errors.

If I delete `callback=parse_comma_str_to_list` from the `data_type` `click-option` it also runs as expected.

Checking how the manifest CLI works, it does not call`callback=parse_comma_str_to_list` to parse this text from the `click-options`. 

I am assuming there is something I am missing here... does it work on your end?

GiaJordan commented:
> @GiaJordan did you experience any issues calling to `validate` after you added this line:
> 
> ```
> @click.option(
>     "-dt",
>     "--data_type",
>     callback=parse_comma_str_to_list,
>     help=query_dict(model_commands, ("model", "validate", "data_type")),
> )
> ```


I have not, but I assume that's because we've been passing in the data type from the command line since it was added.

> When trying to run the command `schematic model --config config.yml validate -mp tests/data/1069/Bulk_rnaseq_filebased_example_error.model.csv` with `data_type` in the `config.yml` file, I get the following error stack:
> 
> ```
> (schematicpy-py3.10) sh-3.2# schematic model --config config.yml validate -mp tests/data/1069/Bulk_rnaseq_filebased_example_error.model.csv
> Starting schematic...
> Traceback (most recent call last):
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/bin/schematic", line 6, in <module>
>     sys.exit(main())
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1257, in invoke
>     sub_ctx = cmd.make_context(cmd_name, args, parent=ctx)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 700, in make_context
>     self.parse_args(ctx, args)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1048, in parse_args
>     value, args = param.handle_parse_result(ctx, opts, args)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 1630, in handle_parse_result
>     value = invoke_param_callback(self.callback, ctx, self, value)
>   File "/Users/mialydefelice/Library/Caches/pypoetry/virtualenvs/schematicpy-UMpwjPmS-py3.10/lib/python3.10/site-packages/click/core.py", line 123, in invoke_param_callback
>     return callback(ctx, param, value)
>   File "/Users/mialydefelice/Documents/schematic_d/schematic/schematic/utils/cli_utils.py", line 165, in parse_comma_str_to_list
>     return comma_string.split(",")
> ```
> 
> When I look what's happening, it looks like the `data_type` isn't being read in at all and `comma_string` here is empty, which is why the error is generated:
> 
> ```
> /Bulk_rnaseq_filebased_example_error.model.csv
> Starting schematic...
> > /Users/mialydefelice/Documents/schematic_d/schematic/schematic/utils/cli_utils.py(165)parse_comma_str_to_list()
> -> return comma_string.split(",")
> (Pdb) comma_string
> (Pdb) param
> <Option data_type>
> ```
> 
> If I add the `-dt` option from the command line it works:
> 
> ```
> Starting schematic...
> 
> error: For the attribute 'Sample ID', on row 3, '' is too short.
> error: For the attribute 'Filename', on row 3, '' is too short.
> [[3, 'Sample ID', "'' is too short", ''], [3, 'Filename', "'' is too short", '']]
> ```
> 
> Returning the expected errors.
> 
> If I delete `callback=parse_comma_str_to_list` from the `data_type` `click-option` it also runs as expected.
> 
> Checking how the manifest CLI works, it does not call`callback=parse_comma_str_to_list` to parse this text from the `click-options`.
> 
> I am assuming there is something I am missing here... does it work on your end?

The callback does get called on my end regardless of whether or not `-dt` is passed in, but I do see the error when it isn't passed in and instead is supplied through the CLI. I've added safeguards now so it can handle leaving the `-dt` parameter blank in the call and gets it from the config when that is the case. Thanks for catching!
@mialy-defelice

mialy-defelice reviewed (approved):
Notes:

- Tested the scenarios that Milen brought up at code review. They passed. 
- Reviewed code prior to review.

Looks great Gianna! Thanks for the change to the CLI :)

GiaJordan commented:
@mialy-defelice 
totally, thanks for your help!
